### PR TITLE
Monthly report with hangfire and min.io implementation

### PR DESCRIPTION
**Pull request by Mary0110**

### CODE REVIEW

File: Services/Authorization/PsychologicalSupportPlatform.Authorization.API/Controllers/AccountsController.cs [hunk 1]
```diff
@@ -2,38 +2,39 @@
...
-        [HttpPost]
-        [Route("login")]
+        [HttpPost("login")]
         public async Task<IActionResult> LoginAsync(LoginData data)
         {
-            var response = await loginService.GetTokenAsync(data);
+            var response = await _accountsService.GetTokenAsync(data);
 
-            if (response.Data is null) return Unauthorized(response.Message);
+            if (response.Data is null)
```
Comment (katyalys) [inline Services/Authorization/PsychologicalSupportPlatform.Authorization.API/Controllers/AccountsController.cs:25]: read about result pattern. you can define model with the property that will hold the actual data and status property. status property will be enum with different status codes. you will return this model from both service and controller (maybe in controller you will need to use ToActionResult() method to return such model). and property with status code will eliminate the need of all checks and returns of Ok, Unauthorized and etc. in contoller.

File: Services/Meet/PsychologicalSupportPlatform.Meet.Application/Meetups/Commands/OrderMeetupByPsychologist/OrderMeetupByPsychologistCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    public OrderMeetupByPsychologistCommandHandler(IMeetupRepository meetupRepository, IScheduleCellRepository scheduleCellRepository, IMapper mapper)
+    {
+        _meetupRepository = meetupRepository;
+        _scheduleCellRepository = scheduleCellRepository;
+        _mapper = mapper;
+    }
+
+    public async Task<int> Handle(OrderMeetupByPsychologistCommand request, CancellationToken cancellationToken)
+    {
+        var chosenScheduleCell = await _scheduleCellRepository.GetByIdAsync(request.MeetupDto.ScheduleCellId);
```
Comment (katyalys) [inline Services/Meet/PsychologicalSupportPlatform.Meet.Application/Meetups/Commands/OrderMeetupByPsychologist/OrderMeetupByPsychologistCommandHandler.cs:24]: move all possible checks to validator as we discussed

**Review by katyalys — CHANGES_REQUESTED** [review]

File: Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs [hunk 3]
```diff
@@ -18,53 +18,54 @@ namespace PsychologicalSupportPlatform.Meet.API.Controllers
...
+            var response = await _mediator.Send(command);
 
             return Ok(response);
         }
 
-        [HttpPut]
+        [HttpPut("{id}")]
         [Authorize(Roles = Roles.Admin + "," + Roles.Psychologist)]
-        public async Task<IActionResult> UpdateMeetup(MeetupDTO meetup)
+        public async Task<IActionResult> UpdateMeetup([FromRoute] int id, AddMeetupDTO meetup)
```
[the remark this answers is not in the document]
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs:65]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs [hunk 4]
```diff
@@ -74,48 +75,48 @@ public async Task<IActionResult> UpdateMeetup(MeetupDTO meetup)
...
         {
             var command = new GetMeetupsByStudentIdQuery(studentId, pageNumber, pageSize);
-            var response = await mediator.Send(command);
+            var response = await _mediator.Send(command);
 
             return Ok(response);
         }
 
-        [HttpPut("approve/meetup_id={meetupId}")]
+        [HttpPut("student/meetup-approval/{meetupId}")]
```
[the remark this answers is not in the document]
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs:113]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs [hunk 5]
```diff
@@ -74,48 +75,48 @@ public async Task<IActionResult> UpdateMeetup(MeetupDTO meetup)
...
         {
             var command = new GetMeetupsByDateQuery(date, pageNumber, pageSize);
-            var response = await mediator.Send(command);
+            var response = await _mediator.Send(command);
 
             return Ok(response);
         }
         
-        [HttpGet("student={studentId}")]
+        [HttpGet("students/{studentId}/meetups")]
```
[the remark this answers is not in the document]
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs:103]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/ScheduleCellsController.cs [hunk 6]
```diff
@@ -15,43 +15,42 @@ namespace PsychologicalSupportPlatform.Meet.API.Controllers
...
         
         [HttpPost]
         [Authorize(Roles = Roles.Admin + "," + Roles.Psychologist)]
         public async Task<IActionResult> CreateScheduleCell(CreateScheduleCellDTO scheduleCell)
         {
-            var createCmdScheduleCell = mapper.Map<AddScheduleCellDTO>(scheduleCell);
-            var command = mapper.Map<CreateScheduleCellCommand>(createCmdScheduleCell);
-            var response = await mediator.Send(command);
+            var createCmdScheduleCell = _mapper.Map<AddScheduleCellDTO>(scheduleCell);
+            var command = _mapper.Map<CreateScheduleCellCommand>(createCmdScheduleCell);
```
[the remark this answers is not in the document]
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/ScheduleCellsController.cs:32]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.Application/Meetups/Commands/Update/UpdateMeetupCommand.cs [hunk 7]
```diff
@@ -1,7 +1,8 @@
 using MediatR;
 using PsychologicalSupportPlatform.Common;
 using PsychologicalSupportPlatform.Meet.Application.DTOs;
+using PsychologicalSupportPlatform.Meet.Application.DTOs.Meetup;
 
 namespace PsychologicalSupportPlatform.Meet.Application.Meetups.Commands.Update;
 
-public record UpdateMeetupCommand(MeetupDTO MeetupDTO): IRequest<int>;
+public record UpdateMeetupCommand(int id, AddMeetupDTO MeetupDTO): IRequest<int>;
```
[the remark this answers is not in the document]
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.Application/Meetups/Commands/Update/UpdateMeetupCommand.cs:8]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.Application/ScheduleCells/Commands/Update/UpdateScheduleCellCommand.cs [hunk 8]
```diff
@@ -1,7 +1,6 @@
 using MediatR;
-using PsychologicalSupportPlatform.Common;
-using PsychologicalSupportPlatform.Meet.Application.DTOs;
+using PsychologicalSupportPlatform.Meet.Application.DTOs.ScheduleCell;
 
 namespace PsychologicalSupportPlatform.Meet.Application.ScheduleCells.Commands.Update;
 
-public record UpdateScheduleCellCommand(ScheduleCellDTO scheduleCellDTO): IRequest<int>;
+public record UpdateScheduleCellCommand(int id, AddScheduleCellDTO scheduleCellDTO): IRequest<int>;
```
[the remark this answers is not in the document]
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.Application/ScheduleCells/Commands/Update/UpdateScheduleCellCommand.cs:6]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.Infrastructure/Data/Repositories/ScheduleCellRepository.cs [hunk 9]
```diff
@@ -19,9 +19,10 @@ public async Task<List<ScheduleCell>> GetScheduleCellsByStatusAsync(bool status,
         return await GetAllAsync(cell => cell.Active == status, pageNumber, pageSize);
     }
 
-    public async Task<List<ScheduleCell>> GetScheduleCellsByDayAndTimeAsync(DayOfWeek dayOfWeek, TimeOnly time)
+    public async Task<List<ScheduleCell>> GetScheduleCellsByDayAndTimeAsync(DayOfWeek dayOfWeek, TimeOnly time, int PsyId)
```
[the remark this answers is not in the document]
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.Infrastructure/Data/Repositories/ScheduleCellRepository.cs:22]: fixed

File: Services/Report/PsychologicalSupportPlatform.Report.API/Program.cs [hunk 10]
```diff
@@ -16,9 +18,15 @@
...
 builder.Services.AddRouting(options => options.LowercaseUrls = true);
 builder.Services.AddRabbitMQBackground(builder.Configuration);
+builder.Services.AddHangfireService(builder.Configuration);
+builder.Services.ConfigureMinio(builder.Configuration);
 
 var app = builder.Build();
 
+app.UseHangfireDashboard();
+
+RecurringJob.AddOrUpdate<IMonthlyReportService>("monthly-report-job", monthlyReportService => monthlyReportService.AddMonthlyReportAsync(), Cron.Monthly);
```
[the remark this answers is not in the document]
Comment (Mary0110) [reply Services/Report/PsychologicalSupportPlatform.Report.API/Program.cs:28]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.Application/Meetups/Commands/OrderMeetupByPsychologist/OrderMeetupByPsychologistCommandHandler.cs (hunk 2 above)
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.Application/Meetups/Commands/OrderMeetupByPsychologist/OrderMeetupByPsychologistCommandHandler.cs:24]: fixed

File: Services/Authorization/PsychologicalSupportPlatform.Authorization.API/Controllers/AccountsController.cs (hunk 1 above)
Comment (Mary0110) [reply Services/Authorization/PsychologicalSupportPlatform.Authorization.API/Controllers/AccountsController.cs]: Changed ResponseInfo and DataResponseInfo classes

File: Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs [hunk 11]
```diff
@@ -110,7 +110,7 @@ public async Task<IActionResult> GetMeetupByStudentId([FromRoute] int studentId,
             return Ok(response);
         }
 
-        [HttpPut("student/meetup-approval/{meetupId}")]
+        [HttpPut("{meetupId}student/meetup-approval")]
```
Comment (mebry) [inline Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs]: You forgot about /. {meetupId}/student*

File: Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/ScheduleCellsController.cs [hunk 12]
```diff
@@ -28,8 +28,7 @@ public ScheduleCellsController(IMediator mediator, IMapper mapper)
         [Authorize(Roles = Roles.Admin + "," + Roles.Psychologist)]
         public async Task<IActionResult> CreateScheduleCell(CreateScheduleCellDTO scheduleCell)
         {
-            var createCmdScheduleCell = _mapper.Map<AddScheduleCellDTO>(scheduleCell);
-            var command = _mapper.Map<CreateScheduleCellCommand>(createCmdScheduleCell);
+            var command = _mapper.Map<CreateScheduleCellCommand>(scheduleCell);
```
Comment (mebry) [inline Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/ScheduleCellsController.cs:31]: In your commands, you can accept DTO models to eliminate unnecessary mapping in the controllers

File: Services/Messaging/PsychologicalSupportPlatform.Messaging.Application/Validators/AddMessageValidator.cs [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+public class AddMessageValidator: AbstractValidator<AddMessageDTO>
+{
+    public AddMessageValidator()
+    {
+        RuleFor(dto => dto.Text).NotEmpty();
+        RuleFor(dto => dto.ConsumerId).NotEmpty().Must(ValidatorHelper.IsValidId);
+        RuleFor(dto => dto.DateTime).NotEmpty().GreaterThanOrEqualTo(DateTime.Now);
+        RuleFor(dto => dto.SenderId).NotEmpty().Must(ValidatorHelper.IsValidId);
+    }
```
Comment (mebry) [inline Services/Messaging/PsychologicalSupportPlatform.Messaging.Application/Validators/AddMessageValidator.cs:14]: Add an empty line between each rule

File: Services/Messaging/PsychologicalSupportPlatform.Messaging.Application/Validators/GetChatHistoryUserValidator.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace PsychologicalSupportPlatform.Messaging.Application.Validators;
+
+public class GetChatHistoryValidator: AbstractValidator<GetChatHistoryDTO>
+{
+    public GetChatHistoryValidator()
+    {
+        RuleFor(dto => dto.OtherUserId).NotEmpty().Must(ValidatorHelper.IsValidId);
+        RuleFor(dto => dto.SenderId).NotEmpty().Must(ValidatorHelper.IsValidId);
+    }
```
Comment (mebry) [inline Services/Messaging/PsychologicalSupportPlatform.Messaging.Application/Validators/GetChatHistoryUserValidator.cs:12]: The same

**Review by mebry — CHANGES_REQUESTED** [review]

File: Services/Messaging/PsychologicalSupportPlatform.Messaging.Application/Validators/GetChatHistoryUserValidator.cs (hunk 14 above)
Comment (Mary0110) [reply Services/Messaging/PsychologicalSupportPlatform.Messaging.Application/Validators/GetChatHistoryUserValidator.cs:12]: fixed

File: Services/Messaging/PsychologicalSupportPlatform.Messaging.Application/Validators/AddMessageValidator.cs (hunk 13 above)
Comment (Mary0110) [reply Services/Messaging/PsychologicalSupportPlatform.Messaging.Application/Validators/AddMessageValidator.cs:14]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/ScheduleCellsController.cs (hunk 12 above)
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/ScheduleCellsController.cs:31]: fixed

File: Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs (hunk 11 above)
Comment (Mary0110) [reply Services/Meet/PsychologicalSupportPlatform.Meet.API/Controllers/MeetupsController.cs]: fixed

**Review by katyalys — APPROVED** [review]

**Pull request merged**